In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
path = 'https://classes.cs.uoregon.edu/18F/cis199ids/data/FIFA2018Statistics.csv'
fifa = pd.read_csv(path)

In [3]:
fifa.head()

Date          Team      Opponent  Goal Scored  Ball Possession %  \
0  14-06-2018        Russia  Saudi Arabia            5                 40   
1  14-06-2018  Saudi Arabia        Russia            0                 60   
2  15-06-2018         Egypt       Uruguay            0                 43   
3  15-06-2018       Uruguay         Egypt            1                 57   
4  15-06-2018       Morocco          Iran            0                 64   

   Attempts  On-Target  Off-Target  Blocked  Corners      ...        \
0        13          7           3        3        6      ...         
1         6          0           3        3        2      ...         
2         8          3           3        2        0      ...         
3        14          4           6        4        5      ...         
4        13          3           6        4        5      ...         

   Yellow Card  Yellow & Red  Red  Man of the Match  1st Goal        Round  \
0            0             0    0               Yes      12.0  Group Stage   
1            0             0    0                No       NaN  Group Stage   
2            2             0    0                No       NaN  Group Stage   
3            0             0    0               Yes      89.0  Group Stage   
4            1             0    0                No       NaN  Group Stage   

   PSO  Goals in PSO  Own goals  Own goal Time  
0   No             0        NaN            NaN  
1   No             0        NaN            NaN  
2   No             0        NaN            NaN  
3   No             0        NaN            NaN  
4   No             0        1.0           90.0  

[5 rows x 27 columns]

In [4]:
fifa.sample(n=15, random_state=1)

Date         Team        Opponent  Goal Scored  Ball Possession %  \
121  10-07-2018      Belgium          France            0                 60   
104  02-07-2018       Brazil          Mexico            2                 47   
55   23-06-2018       Mexico  Korea Republic            2                 59   
56   23-06-2018      Germany          Sweden            2                 71   
33   19-06-2018        Egypt          Russia            1                 53   
118  07-07-2018       Russia         Croatia            2                 38   
122  11-07-2018      Croatia         England            2                 54   
35   20-06-2018      Morocco        Portugal            0                 53   
31   19-06-2018      Senegal          Poland            2                 43   
51   22-06-2018  Switzerland          Serbia            2                 58   
108  03-07-2018       Sweden     Switzerland            1                 37   
103  01-07-2018      Denmark         Croatia            1                 46   
4    15-06-2018      Morocco            Iran            0                 64   
67   25-06-2018        Egypt    Saudi Arabia            1                 39   
102  01-07-2018      Croatia         Denmark            1                 54   

     Attempts  On-Target  Off-Target  Blocked  Corners      ...        \
121         9          3           5        1        5      ...         
104        21         10           7        4        8      ...         
55         13          5           6        2        5      ...         
56         16          5           4        7        8      ...         
33         13          1           8        4        4      ...         
118        13          7           4        2        6      ...         
122        22          7          11        4        8      ...         
35         16          4          10        2        7      ...         
31          8          2           4        2        3      ...         
51         20          5           8        7        7      ...         
108        12          3           6        3        3      ...         
103        15          3          10        2        4      ...         
4          13          3           6        4        5      ...         
67          8          1           6        1        2      ...         
102        22          7           8        7        5      ...         

     Yellow Card  Yellow & Red  Red  Man of the Match  1st Goal  \
121            3             0    0                No       NaN   
104            2             0    0               Yes      51.0   
55             0             0    0               Yes      26.0   
56             0             1    0               Yes      48.0   
33             1             0    0                No      47.0   
118            1             0    0                No      31.0   
122            2             0    0               Yes      68.0   
35             1             0    0                No       NaN   
31             2             0    0               Yes      60.0   
51             1             0    0               Yes      52.0   
108            1             0    0               Yes      66.0   
103            1             0    0                No       1.0   
4              1             0    0                No       NaN   
67             2             0    0               Yes      22.0   
102            0             0    0               Yes       4.0   

              Round  PSO  Goals in PSO  Own goals  Own goal Time  
121    Semi- Finals   No             0        NaN            NaN  
104     Round of 16   No             0        NaN            NaN  
55      Group Stage   No             0        NaN            NaN  
56      Group Stage   No             0        NaN            NaN  
33      Group Stage   No             0        1.0           47.0  
118  Quarter Finals  Yes             3        NaN            NaN  
122    Semi- Finals

In [5]:
for item in fifa.columns:
    print(item)

Date
Team
Opponent
Goal Scored
Ball Possession %
Attempts
On-Target
Off-Target
Blocked
Corners
Offsides
Free Kicks
Saves
Pass Accuracy %
Passes
Distance Covered (Kms)
Fouls Committed
Yellow Card
Yellow & Red
Red
Man of the Match
1st Goal
Round
PSO
Goals in PSO
Own goals
Own goal Time


There are multiple columns you could drop but if I had to choose 2 I would probably choose:
    1. Date
    2. Distance
I would choose these because they do not factor in to determining the 'Man of the Match'

In [6]:
X_fifa = fifa.drop('Man of the Match', axis=1)

In [7]:
y_fifa = fifa['Man of the Match']

In [8]:
fifa['Team'].unique()

array(['Russia', 'Saudi Arabia', 'Egypt', 'Uruguay', 'Morocco', 'Iran',
       'Portugal', 'Spain', 'France', 'Australia', 'Argentina', 'Iceland',
       'Peru', 'Denmark', 'Croatia', 'Nigeria', 'Costa Rica', 'Serbia',
       'Germany', 'Mexico', 'Brazil', 'Switzerland', 'Sweden',
       'Korea Republic', 'Belgium', 'Panama', 'Tunisia', 'England',
       'Colombia', 'Japan', 'Poland', 'Senegal'], dtype=object)

In [9]:
fifa['Opponent'].unique()

array(['Saudi Arabia', 'Russia', 'Uruguay', 'Egypt', 'Iran', 'Morocco',
       'Spain', 'Portugal', 'Australia', 'France', 'Iceland', 'Argentina',
       'Denmark', 'Peru', 'Nigeria', 'Croatia', 'Serbia', 'Costa Rica',
       'Mexico', 'Germany', 'Switzerland', 'Brazil', 'Korea Republic',
       'Sweden', 'Panama', 'Belgium', 'England', 'Tunisia', 'Japan',
       'Colombia', 'Senegal', 'Poland'], dtype=object)

These will not work in our model because they are strings and not actual values, like ints or T/F.

In [10]:
X_fifa = pd.get_dummies(X_fifa, columns=['Team', 'Opponent'])
X_fifa.head()

Date  Goal Scored  Ball Possession %  Attempts  On-Target  \
0  14-06-2018            5                 40        13          7   
1  14-06-2018            0                 60         6          0   
2  15-06-2018            0                 43         8          3   
3  15-06-2018            1                 57        14          4   
4  15-06-2018            0                 64        13          3   

   Off-Target  Blocked  Corners  Offsides  Free Kicks       ...         \
0           3        3        6         3          11       ...          
1           3        3        2         1          25       ...          
2           3        2        0         1           7       ...          
3           6        4        5         1          13       ...          
4           6        4        5         0          14       ...          

   Opponent_Portugal  Opponent_Russia  Opponent_Saudi Arabia  \
0                  0                0                      1   
1                  0                1                      0   
2                  0                0                      0   
3                  0                0                      0   
4                  0                0                      0   

   Opponent_Senegal  Opponent_Serbia  Opponent_Spain  Opponent_Sweden  \
0                 0                0               0                0   
1                 0                0               0                0   
2                 0                0               0                0   
3                 0                0               0                0   
4                 0                0               0                0   

   Opponent_Switzerland  Opponent_Tunisia Opponent_Uruguay  
0                     0                 0                0  
1                     0                 0                0  
2                     0                 0                1  
3                     0                 0                0  
4                     0                 0                0  

[5 rows x 88 columns]

In [11]:
X_fifa.shape, y_fifa.shape

((128, 88), (128,))

The rows need to match because thats how many games were played, the columns shouldn't match because we have multiple features but only one target.

In [12]:
from sklearn.model_selection import train_test_split

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X_fifa, y_fifa, test_size = .30, random_state = 199)

In [14]:
X_train = X_train.copy()
X_train = X_train.drop(['Date', 'PSO', 'Goals in PSO', 'Round', 'Own goals', 'Own goal Time'], axis=1)
X_tets = X_test.copy()
X_test = X_test.drop(['Date', 'PSO', 'Goals in PSO', 'Round', 'Own goals', 'Own goal Time'], axis=1)

In [15]:
X_train['1st Goal'] = X_train['1st Goal'].fillna(0)
X_test['1st Goal'] = X_test['1st Goal'].fillna(0)

Yes, filling in 1st goal with a 0 instead of NaN is good because if the value is NaN then that means the team did not score a goal that particular match. Filling it in with the average or the median would give bad predictions and bad accuracy.

In [16]:
from sklearn.naive_bayes import GaussianNB

In [17]:
model = GaussianNB( )

In [18]:
model.fit(X_train, y_train)

GaussianNB(priors=None)

In [19]:
y_predictions = model.predict(X_test)

In [20]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_predictions)

0.5641025641025641

An accuracy score of 56.4% is not very good at all. Some things we could do to improve our accuracy score would be taking out more feature data that is not helpful, such as 'Team' or 'Opponent'.

In [21]:
from sklearn.tree import DecisionTreeClassifier

In [22]:
model = DecisionTreeClassifier(max_leaf_nodes=8,
                               max_depth=5,
                               min_samples_leaf=20,
                               criterion='gini',
                               splitter='random',
                               random_state=199)

model.fit(X_train, y_train);

y_predictions = model.predict(X_test)

accuracy_score(y_test, y_predictions)

0.6153846153846154

In [23]:
from sklearn.tree import export_graphviz

dot_data = export_graphviz(model,
                           out_file='fifa.dot',
                           feature_names=X_train.columns,
                           class_names=['Man of the Match', 'No Man of the Match'],
                           rounded=True,
                           filled=True)

In [24]:
!dot -Tjpg fifa.dot -o my-fifa-decision-tree14-6-4.jpg

'dot' is not recognized as an internal or external command,
operable program or batch file.
